shared_client.go

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build go1.18
+// +build go1.18
+
+package dns
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// SharedClients holds a set of SharedClient instances.
+type SharedClients struct {
+	sync.Mutex
+	// clients are created and destroyed on demand, hence 'Mutex' needs to be taken.
+	clients map[string]*SharedClient
+}
+
+func NewSharedClients() *SharedClients {
+	return &SharedClients{
+		clients: make(map[string]*SharedClient),
+	}
+}
+
+// GetSharedClient gets or creates an instance of SharedClient keyed with 'key'.
+// if 'key' is an empty sting, a new client is always created and it is not actually shared.
+// The returned 'closer' must be called once the client is no longer needed.
+func (s *SharedClients) GetSharedClient(key string, conf *Client, serverAddrStr string) (client *SharedClient, closer func()) {
+	s.Lock()
+	defer s.Unlock()
+
+	if key != "" {
+		// locate client to re-use if possible.
+		client = s.clients[key]
+	}
+	if client == nil {
+		client = newSharedClient(conf, serverAddrStr)
+		if key != "" {
+			s.clients[key] = client
+		}
+	}
+	client.refcount++
+
+	return client, func() {
+		s.Lock()
+		defer s.Unlock()
+		client.refcount--
+		if client.refcount == 0 {
+			// Make client unreachable and close it's connection.
+			// Must hold the proxy mutex for this.
```

Comment:
Nitpick: I guess `proxy` is not relevant once this code landed in the library, right?

Reply:
Right, will amend!

Reply:
Done, thanks for the review!